internal/lang/funcs/cidr_test.go

```diff
@@ -396,3 +396,107 @@ func TestCidrSubnets(t *testing.T) {
 		})
 	}
 }
+
+func TestCidrContains(t *testing.T) {
+	noError := func(err error) bool { return err == nil }
+
+	tests := []struct {
+		Prefix  cty.Value
+		Address cty.Value
+		Want    cty.Value
+		ErrFn   func(error) bool
+	}{
+		{
+			// IPv4, contained (IP).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.168.2.1"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv4, contained (CIDR).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.168.2.0/22"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv4, not contained.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.126.2.1"),
+			cty.False,
+			noError,
+		},
+		{
+			// IPv4, not contained (CIDR).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.126.2.0/18"),
+			cty.False,
+			noError,
+		},
+		{
+			// IPv6, contained.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("fe80::1"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv6, not contained.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("fe81::1"),
+			cty.False,
+			noError,
+		},
+		{
+			// Address family mismatch: IPv4 prefix, IPv6 address.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("fe80::1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "address family mismatch: 192.168.2.0/20 vs. fe80::1"
+			},
+		},
+		{
+			// Address family mismatch: IPv6 prefix, IPv4 address.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("192.168.2.1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "address family mismatch: fe80::/48 vs. 192.168.2.1"
+			},
+		},
+		{
+			// Input error: invalid CIDR address.
+			cty.StringVal("not-a-cidr"),
+			cty.StringVal("192.168.2.1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "invalid CIDR address: not-a-cidr"
+			},
+		},
+		{
+			// Input error: invalid IP address.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("not-an-address"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "invalid IP address or prefix: not-an-address"
+			},
+		},
+	}
+
```

Comment:
In spelunking through ipaddr today, I reviewed [its readme](https://github.com/opentffoundation/opentf/blob/main/internal/ipaddr/README.md).  In summary, it notes that ipaddr is forked from go 1.16 to preserve behaviour around addresses like '10.010.020.1'.  Go 1.16 trimmed the 0s and went on; Go 1.17 treats the leading 0 as octal digits.  The Go 1.17 behaviour is more popular (and traditional) behaviour.

Should we add a test for this behaviour as well?

Here, I don't think we should add a test.  To add a test would imply we have a preference.  I think holding such a preference would be wrong; we would be propagating a behaviour that's well-contained in one section of code.  

Additionally, should the project ever decide to move past the forked net/ipaddr behaviour, we'd be making additional work.